aptos-move/block-executor/src/view.rs

```diff
@@ -700,57 +756,54 @@ impl<'a, T: Transaction> ResourceGroupState<T> for ParallelState<'a, T> {
             ) {
                 Ok((version, value_with_layout)) => {
                     // If we have a known layout, upgrade RawFromStorage value to Exchanged.
-                    match value_with_layout {
-                        ValueWithLayout::RawFromStorage(v) => {
-                            let patched_value = patch_base_value(v.as_ref(), maybe_layout)?;
+                    if let UnknownOrLayout::Known(layout) = layout {
+                        if let ValueWithLayout::RawFromStorage(v) = value_with_layout {
+                            assert_eq!(version, Err(StorageVersion),
+                            "Fetched resource has unknown layout but the version is not Err(StorageVersion)"
+                            );
+                            let patched_value = patch_base_value(v.as_ref(), layout)?;
+
                             self.versioned_map
                                 .group_data()
                                 .update_tagged_base_value_with_layout(
                                     group_key.clone(),
                                     resource_tag.clone(),
                                     patched_value,
-                                    maybe_layout.cloned().map(Arc::new),
+                                    layout.cloned().map(Arc::new),
```

Comment:
Is this a layout deep clone? We really should change that data structure to at least use arcs...